modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java

```diff
@@ -2653,6 +2766,169 @@ private void _validate(
 		}
 	}
 
+	private String _validateJSON(
+			CPDefinition cpDefinition, CPInstance cpInstance, String json)
+		throws PortalException {
+
+		JSONArray sanitizedJSONArray = _jsonFactory.createJSONArray();
+
+		if (cpDefinition != null) {
+			List<CPDefinitionOptionRel> cpDefinitionOptionRels =
+				_cpDefinitionOptionRelLocalService.getCPDefinitionOptionRels(
+					cpDefinition.getCPDefinitionId());
+			JSONObject jsonObject;
+			JSONArray originalJSONArray = CPJSONUtil.toJSONArray(json);
+
+			for (CPDefinitionOptionRel cpDefinitionOptionRel :
+					cpDefinitionOptionRels) {
+
+				jsonObject = _findMatchingJSONValue(
+					cpDefinitionOptionRel, originalJSONArray);
+
+				if ((cpDefinitionOptionRel.isRequired() ||
+					 cpDefinitionOptionRel.isSkuContributor()) &&
+					(jsonObject == null)) {
+
+					throw new CPDefinitionOptionRelException(
+						"Required option is missing");
+				}
+
+				if (jsonObject != null) {
+					_validateOriginalJSON(cpDefinitionOptionRel, jsonObject);
+				}
+
+				List<CPDefinitionOptionValueRel> cpDefinitionOptionValueRels =
+					_cpDefinitionOptionValueRelLocalService.
+						getCPDefinitionOptionValueRels(
+							cpDefinitionOptionRel.getCPDefinitionOptionRelId());
+
+				if (!cpDefinitionOptionValueRels.isEmpty()) {
+					for (CPDefinitionOptionValueRel cpDefinitionOptionValueRel :
+							cpDefinitionOptionValueRels) {
+
+						if (Validator.isNotNull(
+								cpDefinitionOptionValueRel.
+									getCPInstanceUuid())) {
+
+							jsonObject = _sanitizeJSON(
+								cpDefinitionOptionRel,
+								cpDefinitionOptionValueRel, cpInstance,
+								jsonObject);
```

Comment:
We should only write the value selected by the user, not all the values. This can be done only with SKU contributor options. Required options should throw an exception if the value is not specified and non-required options shouldn't be sanitized at all, just validated if there is a specified value.